Make language in ECN validation clearer (#3777)
quicwg/base-drafts: draft-ietf-quic-transport.md

#### Receiving ACK Frames {#ecn-ack}

Erroneous application of ECN marks in the network can result in degraded
connection performance.  An endpoint that receives an ACK frame with ECN

> counts therefore validates the counts before using them. An endpoint validates
> these counts by comparing newly received counts against those from the last

Reviewer: ```suggestion
counts therefore validates the counts before using them. It performs this
validation by comparing newly received counts against those from the last
```

#### Receiving ACK Frames {#ecn-ack}

> An endpoint that sets ECT(0) or ECT(1) codepoints on packets records the ECN
> counts it has received in ACK frames. When processing an ACK frame, the
> endpoint MUST validate the increase in ECN counts based on the markings that
> were applied to packets that are newly acknowledged.

Reviewer: ```suggestion
An endpoint that sets ECT(0) or ECT(1) codepoints on packets records the ECN
counts it receives in ACK frames. When processing an ACK frame, the endpoint
MUST validate any increases from previously received ECN counts, as follows:
```
... and drop the line below about ECN validation.

> 
> ECN validation involves the following checks:

Reviewer: ```suggestion
```

* If this ACK frame newly acknowledges a packet that the endpoint sent with
  either ECT(0) or ECT(1) codepoints set, ECN validation fails if ECN counts
  are not present in the ACK frame.  This step protects against both a network
  element that zeroes out ECN bits and a peer that is unable to access ECN
  markings, since the peer could respond without ECN feedback in either case.

* ECN validation fails if the sum of the increases to ECT(0), ECT(1), and CE
  counts in the ACK frame are less than the number of newly acknowledged
  packets that were sent with an ECT codepoint.  This step detects any

>   network remarking from ECT(0), ECT(1), or CE codepoints to Not-ECT.

Reviewer: ```suggestion
  erroneous network overwriting from ECT(0), ECT(1), or CE codepoints to
  Not-ECT.
```

* ECN validation fails if any increase in either ECT(0) or ECT(1) counts, plus
  any increase in the CE count, exceeds the number of packets sent with the
  corresponding ECT codepoint that are newly acknowledged in this ACK frame.

>   This step detects any erroneous network remarking from ECT(0) to ECT(1) (or
>   vice versa).

Reviewer: ```suggestion
  This step detects any erroneous network overwriting from ECT(0) to ECT(1) (or
  vice versa).
```

#### Receiving ACK Frames {#ecn-ack}

> Erroneous application of markings in the network can result in ECN marking
> providing bad information.  Before using ECN counts, an endpoint validates the
> ECN counts by comparing the counts in each ACK frame it processes against the
> last ACK frame that was successfully processed.  The increase in ECN counts is
> validated based on the markings that were applied to packets that are newly
> acknowledged in the ACK frame.

Reviewer: ```suggestion
Erroneous application of ECN marks in the network can result in degraded
connection performance.  An endpoint that receives an ACK frame with ECN
counts therefore validates the counts before using them. An endpoint validates
these counts by comparing newly received ones against those from the last
successfully processed ACK frame. Any increase in ECN counts is validated
based on the markings that were applied to packets that are newly
acknowledged in the ACK frame.
```

If an ACK frame newly acknowledges a packet that the endpoint sent with either
ECT(0) or ECT(1) codepoints set, ECN validation fails if ECN counts are not
present in the ACK frame.  This check detects a network element that zeroes out
ECN bits or a peer that is unable to access ECN markings.

ECN validation fails if the sum of the increase in ECT(0) and ECN-CE counts is
less than the number of newly acknowledged packets that were originally sent

> with an ECT(0) marking.  Similarly, if the sum of the increases to ECT(1) and

Reviewer: "Similarly, ECN validation fails if ..."

ECN-CE counts is less than the number of newly acknowledged packets sent with
an ECT(1) marking.  These checks can detect removal of ECN markings in the
network.

An endpoint could miss acknowledgements for a packet when ACK frames are lost.
It is therefore possible for the total increase in ECT(0), ECT(1), and ECN-CE
counts to be greater than the number of packets acknowledged in an ACK frame.
This is why counts are permitted to be larger than might be accounted for by

> newly acknowledged packets.

Reviewer: ```suggestion
newly acknowledged packets. When this happens, and if validation succeeds, the
local reference counts MUST be increased to match the counts in the ACK frame.
```

We still want this bit, right? Ah, I guess we don't need this because the validation is against the counts from the last processed ACK frame. That works for me. Does this make this a design change?

Author: Yep.  We never really said what value you save, but implied that you took the value from the last ACK.  This simplifies and says "last ACK".  I think that's non-design.

ECN validation MAY fail if the total count for an ECT(0) or ECT(1) marking
exceeds the total number of packets sent with the corresponding marking.   In
particular, an endpoint that never applies a particular marking can fail
validation when a non-zero count for the corresponding marking is received.
This check can detect when packets are marked ECT(0) or ECT(1) in the network.

Processing ECN counts out of order can result in validation failure.  An

> endpoint SHOULD skip ECN validation when an ACK frame does not increase the

Reviewer: ```suggestion
endpoint SHOULD skip ECN validation on an ACK frame that does not increase the
```

ECN validation involves the following checks:

* If this ACK frame newly acknowledges a packet that the endpoint sent with
  either ECT(0) or ECT(1) codepoints set, ECN validation fails if ECN counts
  are not present in the ACK frame.  This step protects against both a network
  element that zeroes out ECN bits and a peer that is unable to access ECN
  markings, since the peer could respond without ECN feedback in either case.

* ECN validation fails if the sum of the increases to ECT(0), ECT(1), and CE
  counts in the ACK frame are less than the number of newly acknowledged
  packets that were sent with an ECT codepoint.  This step detects any

>   network remarking from ECT(0), ECT(1), or CE codepoints to Not-ECT.

Reviewer 2: As you use the word "erroneous" below you should also use it here, however, effectively it unknown if this is intentional or not. Most cases we observes it old boxes that still set the whole byte as ToS field. I would actually also recommend to use the term "overwriting" instead of "remarking" as that probably more indicates that, not matter why this happens, it shouldn't.
